### PR TITLE
Add operators subscopes

### LINKED ISSUE #188: Distinguishing from comparison operators
I've noticed that these two rules use the same grouping:

```cson
  {
    'match': '(?<!\\.)\\b(delete|in|of|instanceof|new|typeof|void)(?!\\s*:)\\b'
    'name': 'keyword.operator.js'
  }
  {
    'match': '!=|!==|<=|>=|<<=|>>=|>>>=|\\*=|(?<!\\()/=|%=|\\+=|\\-=|&=|\\^=|!|%|&|\\*|\\-\\-|\\-|\\+\\+|\\+|~|===|==|=|<>|<|>|!|&&|\\|\\||\\?|\\:|\\^'
    'comment': 'match 2-character operator first'
    'name': 'keyword.operator.js'
  }
```

By using the same groups for both, we can't have syntax highlighting be applied to keywords like `delete` and `typeof`.  What would be the best way to distinguish these?  Maybe remove the `keyword` class from the second group of operators?

### PR DESCRIPTION
**Pull request by MaximSokolov**
Closes #188 
Supersedes and closes #222

Now sure about `<>` operator

It's not correct for some cases:
```js
// ':' isn't `keyword.operator.conditional.ternary`
var obj = { a: "b"}
```
```js
// e.g. '+' operator can be used for concatenation
// it's `keyword.operator.arithmetic`
"str1" + "str2"
// `var1`, `var2` can be anything
var1 + var2
```

### CODE REVIEW

File: grammars/javascript.cson [hunk 1]
```diff
@@ -461,7 +463,7 @@
   }
   {
     'match': '(?<!\\.)\\b(delete|in|of|instanceof|new|typeof|void)(?!\\s*:)\\b'
-    'name': 'keyword.operator.js'
+    'name': 'keyword.operator.$1.js'
```
Comment (50Wliu) [inline grammars/javascript.cson]: Can I have your reasoning behind this change?

File: grammars/javascript.cson [hunk 2]
```diff
@@ -544,9 +546,44 @@
     ]
   }
   {
-    'match': '!=|!==|<=|>=|<<=|>>=|>>>=|\\*=|(?<!\\()/=|%=|\\+=|\\-=|&=|\\^=|!|%|&|\\*|/|\\-\\-|\\-|\\+\\+|\\+|~|===|==|=|<>|<|>|!|&&|\\|\\||\\?|\\:|\\^'
-    'comment': 'match 2-character operator first'
-    'name': 'keyword.operator.js'
+    'match': '%=|\\+=|\\-=|\\*=|(?<!\\()/='
+    'name': 'keyword.operator.assignment.augmented.js'
```
Comment (50Wliu) [inline grammars/javascript.cson]: `augmented` sounds really weird to me :/
Anyone have suggestions for a different name?

**Comment by MaximSokolov** [comment]
Changed:
-  Revert `?`, `:` to `keyword.operator`
- `keyword.operator.assignment` for `:` (when it is assignment of course)

> `augmented` sounds really weird to me :/
Anyone have suggestions for a different name?

[Python](https://github.com/atom/language-python/blob/44e1d1836675a9b19928a5b2cfce83613fd5b706/grammars/python.cson#L132-L135), [Ruby](https://github.com/atom/language-ruby/blob/4b59ca4d75be5bdefebb080274cb7ee14f3c51b3/grammars/ruby.cson#L1942-L1945) use `assignment.augmented` I guess it means assignment augmented by another operator.

File: grammars/javascript.cson (hunk 1 above)
Comment (MaximSokolov) [inline grammars/javascript.cson]: It makes possible to change highlighting for each of these operators.

**Comment by MaximSokolov** [comment]
Should I revert 'keyword.operator.arithmetic' -> 'keyword.operator' for `*`, `/`, `-`, `%`, `+` operators?

**Comment by Victorystick** [comment]
@MaximSokolov That sounds like a good idea. We can't be sure that a `+` is used for arithmetic or concatenation. :)

**Comment by 50Wliu** [comment]
@MaximSokolov Go for it.

**Comment by MaximSokolov** [comment]
Is `keyword.operator.assignment.compound` ok for `%=`, `+=`, `-=`, `*=` ... ?
(instead of `...assignment.augmented`)

**Comment by 50Wliu** [comment]
Sounds better to me :laughing:.
/cc @atom/feedback

**Comment by mehcode** [comment]
As a (self proclaimed) language designer, *compound* is more correct (as you are joining two distinct operators into a new operator, now composed of two parts). People started calling it *augmented* with scripting languages (C, Java, etc. always called it *compound*).

**Comment by MaximSokolov** [comment]
I think it's ready

File: grammars/javascript.cson [hunk 3]
```diff
@@ -857,8 +859,43 @@
...
+      {
+        'match': '&&|!|\\|\\|'
+        'name': 'keyword.operator.logical.js'
+      }
+      {
+        'match': '&|\\||\\^|~'
+        'name': 'keyword.operator.bitwise.js'
+      }
+      {
+        'match': '\\?|\\:'
```
Comment (50Wliu) [inline grammars/javascript.cson]: : doesn't need to be escaped

File: grammars/javascript.cson [hunk 4]
```diff
@@ -857,8 +859,43 @@
...
+        'name': 'keyword.operator.js'
+      }
+      {
+        'match': '\\='
```
Comment (50Wliu) [inline grammars/javascript.cson]: = doesn't need to be escaped

File: grammars/javascript.cson [hunk 5]
```diff
@@ -857,8 +859,43 @@
...
+        'name': 'keyword.operator.assignment.js'
+      }
+      {
+        'match': '\\-\\-'
```
Comment (50Wliu) [inline grammars/javascript.cson]: -- doesn't need to be escaped

File: grammars/javascript.cson [hunk 6]
```diff
@@ -857,8 +859,43 @@
...
+        'name': 'keyword.operator.decrement.js'
+      }
+      {
+        'match': '\\+\\+'
+        'name': 'keyword.operator.increment.js'
+      }
+      {
+        'match': '%|\\*|/|\\-|\\+'
```
Comment (50Wliu) [inline grammars/javascript.cson]: - doesn't need to be escaped

File: spec/javascript-spec.coffee [hunk 7]
```diff
@@ -196,65 +196,108 @@ describe "Javascript grammar", ->
...
         expect(tokens[0]).toEqual value: 'test ', scopes: ['source.js']
-        expect(tokens[1]).toEqual value: '/=', scopes: ['source.js', 'keyword.operator.js']
-        expect(tokens[2]).toEqual value: ' ', scopes: ['source.js']
-        expect(tokens[3]).toEqual value: '2', scopes: ['source.js', 'constant.numeric.js']
+        expect(tokens[1]).toEqual value: '?', scopes: ['source.js', 'keyword.operator.js']
+        expect(tokens[2]).toEqual value: ' expr1 ', scopes: ['source.js']
+        expect(tokens[3]).toEqual value: ':', scopes: ['source.js', 'keyword.operator.js']
+        expect(tokens[4]).toEqual value: ' expr2', scopes: ['source.js']
+
+    describe "logical", ->
```
Comment (50Wliu) [inline spec/javascript-spec.coffee]: describe "logical operators"
and then it "tokenizes them"

Same goes for below

Comment (MaximSokolov) [inline spec/javascript-spec.coffee]: > describe "logical operators"

It's already in `describe "operators"`:
```coffee
describe "operators", ->
  ...
  describe "logical", ->
```

Comment (50Wliu) [inline spec/javascript-spec.coffee]: Oops, didn't notice that, sorry!

File: grammars/javascript.cson [hunk 8]
```diff
@@ -857,8 +859,43 @@
   'operators':
     'patterns': [
       {
-        'match': '!=|!==|<=|>=|<<=|>>=|>>>=|\\*=|(?<!\\()/=|%=|\\+=|\\-=|&=|\\^=|!|%|&|\\*|/|\\-\\-|\\-|\\+\\+|\\+|~|===|==|=|<>|<|>|!|&&|\\|\\||\\?|\\:|\\^'
-        'comment': 'match 2-character operator first'
+        'match': '%=|\\+=|\\-=|\\*=|(?<!\\()/='
```
Comment (50Wliu) [inline grammars/javascript.cson]: Final nit: The - here doesn't need to be escaped either.

File: spec/javascript-spec.coffee (hunk 7 above)
Comment (MaximSokolov) [inline spec/javascript-spec.coffee]: Forget to update `tokenizes them` below. Don't merge it yet

**Comment by MaximSokolov** [comment]
Updated

**Comment by 50Wliu** [comment]
Woo!

**Comment by smaili** [comment]
:+1:

**Pull request merged**